Add sof_dai_get_tdm_slots() API

brentlu opened the pull request:
Machine driver sof_da7219 is using local functions to support max98373 because the slot number is different from the number in maxim-common module. To solve this, we

1. Add an new API sof_dai_get_tdm_slots() to retrieve tdm slot number from topology in runtime.
2. Rewrite max_98373_hw_params() function in maxim-common module to take advantage of the new tdm slot number query API.
3. Remove the local function in sof_da7219 and use the common code in maxim-common to support max98373

brentlu committed:
ASoC: SOF: sof-audio: rename dai clock setting query function

We are going to extend the dai_get_clk() callback function to retrieve
other dai-specific settings like tdm slot number, not just mclk/bclk
frequencies. Revisit functions and macros related to the callback and
replace the string 'clk' with 'param' to reflect the ongoing change.

Signed-off-by: Brent Lu <brent.lu@intel.com>

brentlu committed:
ASoC: SOF: sof-audio: add sof_dai_get_tdm_slots function

An new interface, sof_dai_get_tdm_slots(), is added for machine driver
to get tdm slot number from topology. The dai_get_param() callback
needs to support new parameter type SOF_DAI_PARAM_INTEL_SSP_TDM_SLOTS
by returning the tdm slot number of specific SSP port.

Signed-off-by: Brent Lu <brent.lu@intel.com>

brentlu committed:
ASoC: SOF: ipc3-topology: support tdm slot number query

Support the new parameter type SOF_DAI_PARAM_INTEL_SSP_TDM_SLOTS in
sof_ipc3_dai_get_param() function to get the tdm slot number of
specific SSP port.

Signed-off-by: Brent Lu <brent.lu@intel.com>

brentlu committed:
ASoC: SOF: ipc4-topology: support tdm slot number query

Support the new parameter type SOF_DAI_PARAM_INTEL_SSP_TDM_SLOTS in
sof_ipc4_dai_get_param() function to get the tdm slot number of
specific SSP port.

Signed-off-by: Brent Lu <brent.lu@intel.com>

brentlu committed:
ASoC: Intel: maxim-common: rewrite max_98373_hw_params function

Rewrite max_98373_hw_params() function to get tdm slot number from
topology in runtime because earlier platforms use 4-slot setting
instead of 8. Also check if the interface is configured in TDM mode
before calling snd_soc_dai_set_tdm_slot() function.

Signed-off-by: Brent Lu <brent.lu@intel.com>

brentlu committed:
ASoC: Intel: sof_da7219: remove loacl max98373 ops

Since maxim-module could detect tdm slot number in runtime, we could
remove local ops and use maxim-module to handle max98373.

Signed-off-by: Brent Lu <brent.lu@intel.com>

bardliao commented on sound/soc/intel/boards/sof_maxim_common.c:
```diff
@@ -72,28 +73,61 @@ static struct snd_soc_dai_link_component max_98373_components[] = {
+	for_each_rtd_codec_dais(rtd, i, codec_dai) {
+		if (i >= ARRAY_SIZE(max_98373_tdm_mask)) {
+			dev_err(codec_dai->dev, "invalid codec index %d\n", i);
+			return -ENODEV;
```
Maybe "Currently, only %d amps are supported",  ARRAY_SIZE(max_98373_tdm_mask) is better. And it is more like an EINVAL error.

brentlu replied:
done

brentlu replied:
Type of 'ARRAY_SIZE(max_98373_tdm_mask)' is different under 32 and 64 bit system and results in build error when compiler checks the placeholder %d in string...

plbossart commented on sound/soc/sof/ipc3-topology.c:
```diff
@@ -2509,15 +2509,15 @@ static int sof_ipc3_dai_get_clk(struct snd_sof_dev *sdev, struct snd_sof_dai *da
 			break;
 		}
-		dev_err(sdev->dev, "fail to get SSP clk %d rate\n", clk_type);
+		dev_err(sdev->dev, "fail to get SSP clk %d rate\n", param_type);
```
Invalid SSP param %d

brentlu replied:
fixed

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -3169,13 +3169,13 @@ static int sof_ipc4_dai_get_clk(struct snd_sof_dev *sdev, struct snd_sof_dai *da
 			return le32_to_cpu(hw_config->bclk_rate);
 		default:
-			dev_err(sdev->dev, "Invalid clk type for SSP %d\n", clk_type);
+			dev_err(sdev->dev, "Invalid clk type for SSP %d\n", param_type);
```
Invalid SSP param %d

brentlu replied:
fixed

plbossart commented on sound/soc/intel/boards/sof_maxim_common.c:
```diff
@@ -72,28 +73,61 @@ static struct snd_soc_dai_link_component max_98373_components[] = {
 	struct snd_soc_dai *codec_dai;
-	int ret = 0;
-	int j;
+	int i, tdm_slots, ret = 0;
```
I prefer one line per variable, and certainly initialized variables on a separate line...

brentlu replied:
done

plbossart commented on sound/soc/intel/boards/sof_maxim_common.c:
```diff
@@ -72,28 +73,61 @@ static struct snd_soc_dai_link_component max_98373_components[] = {
+			ret = snd_soc_dai_set_tdm_slot(codec_dai,
+						       max_98373_tdm_mask[i].tx,
+						       max_98373_tdm_mask[i].rx,
+						       tdm_slots,
```
Not following, what happen if tdm_slots is 4, how would this work with the masks set as 0x3?

Is this intended to always transmit two channels no matter what the number of slots might be?

A comment or explanation would help!

brentlu replied:
Yes, max98373 takes two channels as input and let user space to choose the output from one of 'ch0', 'ch1', '(ch0+ch1)/2'. Here the 0x3 tells codec driver to select TDM channel 0 and 1 as input channels.

plbossart commented on sound/soc/intel/boards/sof_maxim_common.c:
```diff
@@ -9,6 +9,7 @@
 #include <sound/soc-acpi.h>
 #include <sound/soc-dai.h>
 #include <sound/soc-dapm.h>
+#include <sound/sof.h>
```
commit message typos: becasue eariler

brentlu replied:
fixed

plbossart replied:
still typo: eariler

plbossart commented on sound/soc/intel/boards/sof_da7219.c:
```diff
@@ -178,42 +178,21 @@ static void da7219_codec_exit(struct snd_soc_pcm_runtime *rtd)
 
-static int max98373_hw_params(struct snd_pcm_substream *substream,
-			      struct snd_pcm_hw_params *params)
+static int card_late_probe(struct snd_soc_card *card)
```
commit message maxin-common -> maxim-module

brentlu replied:
done

plbossart reviewed (changes requested):
Couple of typos and explanations needed @brentlu, otherwise looks good to me.

plbossart commented on sound/soc/intel/boards/sof_maxim_common.c:
```diff
@@ -72,28 +73,71 @@ static struct snd_soc_dai_link_component max_98373_components[] = {
+			break;
 		}
 	}
+
```
spurious line.

brentlu replied:
removed

plbossart commented on sound/soc/intel/boards/sof_maxim_common.c:
```diff
@@ -72,28 +73,71 @@ static struct snd_soc_dai_link_component max_98373_components[] = {
+
+			ret = snd_soc_dai_set_tdm_slot(codec_dai,
+						       max_98373_tdm_mask[i].tx,
+						       max_98373_tdm_mask[i].rx,
```
I would still create a mask of all tx and rx setups, and verify that they are all compatible with the tdm slots.

For example if more than 2 channels are used, and tdm_slot == 2 then something is incorrect.

brentlu replied:
Add checks to make sure tx/rx mask is compatible with tdm slot number.

plbossart commented on sound/soc/intel/boards/sof_da7219.c:
```diff
@@ -178,42 +178,21 @@ static void da7219_codec_exit(struct snd_soc_pcm_runtime *rtd)
+		snd_soc_dapm_disable_pin(dapm, "Right Spk");
+		err = snd_soc_dapm_sync(dapm);
+		if (err < 0)
+			return err;
```
where does this come from? this wasn't done before, not sure why it's part of this patch.

brentlu replied:
It's done in 94d2d0897474 ("ASoC: Intel: Boards: tgl_max98373: add dai_trigger function") on TGL platform. The sof_da7219.c driver was implemented earlier for JSL platform so it's missed.

bardliao replied:
> It's done in [94d2d08](https://github.com/thesofproject/linux/commit/94d2d08974746bd0f469ec5c1e50c613de558bfc) ("ASoC: Intel: Boards: tgl_max98373: add dai_trigger function") on TGL platform. The sof_da7219.c driver was implemented earlier for JSL platform so it's missed.

Please add the missing part with different commit.

brentlu replied:
done

plbossart commented on sound/soc/intel/boards/sof_da7219.c:
```diff
@@ -180,6 +180,19 @@ static void da7219_codec_exit(struct snd_soc_pcm_runtime *rtd)
 
 static int card_late_probe(struct snd_soc_card *card)
 {
+	struct sof_card_private *ctx = snd_soc_card_get_drvdata(card);
```
commit message, please fix the checkpatch error on the SHA1 value

brentlu replied:
fixed

brentlu committed:
ASoC: Intel: sof_da7219: disable max98373 speaker pins in late_probe

Commit 94d2d0897474 ("ASoC: Intel: Boards: tgl_max98373: add dai_trigger
function") disables speaker pins in late_probe to allow DSP to enter
low power state. Add same code to sof_da7219 driver.

Signed-off-by: Brent Lu <brent.lu@intel.com>
